### PR TITLE
Add KL divergence

### PR DESCRIPTION
**Pull request by Nisone2000**
I removed all the hardcoded values, set a seed for kmeans and the trainer will also return the cluster_centers.

### CODE REVIEW

File: notebooks/dataset_tutorial/sensorium/training/trainers.py [hunk 1]
```diff
@@ -40,19 +36,8 @@ def standard_trainer(
     lr_decay_factor=0.3,
     min_lr=0.0001,
     cb=None,
-    use_wandb=True,
-    wandb_project="Rotation_test",
-    wandb_entity="ninasophie-nellen-g-ttingen-university",
```
Comment (pollytur) [inline notebooks/dataset_tutorial/sensorium/training/trainers.py]: I see that its deleted, but as an advice - better not put real project names, etc to git
And NEVER put any keys (like wandb authorisation key to git)

File: notebooks/dataset_tutorial/sensorium/utility/measure_helpers.py [hunk 2]
```diff
@@ -1,49 +1,63 @@
...
 
+
 # smooth color interpolation
 def lerp(x, a, b):
-    return a + x * (b-a)
+    return a + x * (b - a)
+
 
 # smooth color interpolation
 def serp(x, a, b):
```
Comment (pollytur) [inline notebooks/dataset_tutorial/sensorium/utility/measure_helpers.py]: @Nisone2000 I would inspire you to name the functions a bit more explicit as I cannot guess what it is supposed to do from the name:)

File: notebooks/dataset_tutorial/sensorium/utility/measure_helpers.py [hunk 3]
```diff
@@ -76,21 +90,20 @@ def __init__(self, cmap_array, transpose=False, reverse_x=False, reverse_y=False
...
         val = self._scale_to_range(val, xmin, xmax)
-        rescaled = (val * (self._width - 1))
+        rescaled = val * (self._width - 1)
         return rescaled.astype(int)
 
     def _map_to_y(self, val: np.ndarray) -> np.ndarray:
         ymin, ymax = self._range_y
         val = self._scale_to_range(val, ymin, ymax)
-        rescaled = (val * (self._height - 1))
+        rescaled = val * (self._height - 1)
```
Comment (pollytur) [inline notebooks/dataset_tutorial/sensorium/utility/measure_helpers.py]: @Nisone2000 why there is `-1` here? Could you add a comment for it?

File: sensorium/training/trainers.py [hunk 4]
```diff
@@ -68,14 +83,44 @@ def standard_trainer(
         min_lr: minimum learning rate
         cb: whether to execute callback function
         track_training: whether to track and print out the training progress
+
+        cluster_number: Give number of clusters for DEC clustering algortihm
+        alpha: alpha used for calculation of soft assignment
+        dec_starting_epoch: Epoch at which we start the initialisation for the cluster centroids for dec clustering
+        kmeans_init: number of iterations for kmeans for cluster initialisation
```
Comment (pollytur) [inline sensorium/training/trainers.py:91]: nice, thanks for the comments:)

File: sensorium/training/trainers.py [hunk 5]
```diff
@@ -68,14 +83,44 @@ def standard_trainer(
...
 
-    def full_objective(model, dataloader, data_key, *args, **kwargs):
+    def get_multiplier(epoch, base_multiplier=1e8):
+        if epoch < 14:
+            return 0
+        else:
+            return base_multiplier
+
+    def soft_assignments(encoded_features, cluster_centers, alpha=alpha):
+        # Compute soft assingments q_ij as described in DEC paper (1)
```
Comment (pollytur) [inline sensorium/training/trainers.py]: @Nisone2000 maybe add formula (1) here, such that a person who reads it does not have to go to the papers (also that you don't have to go to the paper in case we start debugging it)

File: sensorium/training/trainers.py [hunk 6]
```diff
@@ -84,20 +129,26 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+
+        tot_main_loss = loss_scale * criterion(
+            model(args[0].to(device), data_key=data_key, **kwargs),
+            args[1].to(device),
         )
+        return (tot_main_loss + regularizers), (tot_main_loss, regularizers)
 
     ##### Model training ####################################################################################################
+
+    """
```
Comment (pollytur) [inline sensorium/training/trainers.py:146]: @Nisone2000  what is this? and why is it commented out?

File: sensorium/training/trainers.py [hunk 7]
```diff
@@ -109,7 +160,17 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
 
     n_iterations = len(LongCycler(dataloaders["train"]))
 
-    optimizer = torch.optim.Adam(model.parameters(), lr=lr_init)
+    if include_kldivergence:
+        cluster_centers = torch.zeros(
+            cluster_number, 128, dtype=torch.float, requires_grad=True
```
Comment (pollytur) [inline sensorium/training/trainers.py]: I guess the 128 comes from the dimensionality of the readouts. It might be better if you do do not hardcode it but dynamically assign it based on the model

File: sensorium/training/trainers.py [hunk 8]
```diff
@@ -109,7 +160,17 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+        )
```
Comment (pollytur) [inline sensorium/training/trainers.py]:
```suggestion
        k = list(model.readout.keys())[0]
        dim = model.readout[k].features.shape[1]
        dtype = model.readout[k].features.dtype
        cluster_centers = torch.zeros(
            cluster_number, dim, dtype=dtype, requires_grad=True
        )
```
please make sure that this works!

File: sensorium/training/trainers.py (hunk 6 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py:146]: That's if we use the class DEC as they did it in the paper. They save the cluster centroids as a property of the model s.t. it gets automaticly updated in the optimising step. Right now I'm saving the cluster centroids in a tensor.

File: sensorium/training/trainers.py (hunk 7 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: True. I will fix that

File: sensorium/training/trainers.py [hunk 9]
```diff
@@ -168,6 +252,40 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
         lr_decay_steps=lr_decay_steps,
     ):
 
+        if include_kldivergence and epoch == dec_starting_epoch:
+            # TODO: include hidden dimension
+            kmeans = KMeans(
+                n_clusters=cluster_number, n_init=kmeans_init, random_state=seed
+            )
+            model.train()
```
Comment (pollytur) [inline sensorium/training/trainers.py]: I think this line should be after the things inside this `if` statement
like 
```
if include_kldivergence and epoch == dec_starting_epoch:
    kmeans = KMeans(
                n_clusters=cluster_number, n_init=kmeans_init, random_state=seed
            )
    ...
model.train()
if verbose and tracker is not None:
            print("=======================================")
```
Please correct me if I am wrong

File: sensorium/training/trainers.py [hunk 10]
```diff
@@ -168,6 +252,40 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+                features = readout.features.cpu().detach().squeeze().T.numpy()
+                feature_list.append(np.array(features))
+                random_index = np.random.choice(
+                    features.shape[0], size=int(features.shape[0] / 5), replace=False
+                )
+                random_indices[k] = random_index
+                features_subset.append(features[random_indices[k], :])
+            features = np.vstack(feature_list)
+            features_subset = np.vstack(features_subset)
+            predicted = kmeans.fit_predict(features)
```
Comment (pollytur) [inline sensorium/training/trainers.py]: I think this is initialization, so there should be no gradients through this stage
To make it more explicit you can do sth like
```suggestion
            with torch.no_grad():
                for k, readout in model.readout.items():
                    features = readout.features.cpu().detach().squeeze().T.numpy()
                    feature_list.append(np.array(features))
                    random_index = np.random.choice(
                        features.shape[0], size=int(features.shape[0] / 5), replace=False
                    )
                    random_indices[k] = random_index
                    features_subset.append(features[random_indices[k], :])
                features = np.vstack(feature_list)
                features_subset = np.vstack(features_subset)
                predicted = kmeans.fit_predict(features)
```

File: sensorium/training/trainers.py [hunk 11]
```diff
@@ -188,21 +311,116 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+            epoch_loss_reg += loss_parts[1].detach()
+            cluster_centers_list = []
             if (batch_no + 1) % optim_step_count == 0:
+                # TODO maybe remove the hidden dimensions
+                if include_kldivergence and epoch >= dec_starting_epoch:
+                    kldiv_loss = torch.zeros(1).to(device)
+                    features_subset = []
+                    feature_list = []
+                    for k, readout in model.readout.items():
+                        features = readout.features.detach().squeeze()
```
Comment (pollytur) [inline sensorium/training/trainers.py]: why do you do `detach` here?
I would suggest to do `features = readout.features.squeeze()` because otherwise this might mess up gradient [computations](https://pytorch.org/docs/stable/generated/torch.Tensor.detach.html#)

File: sensorium/training/trainers.py [hunk 12]
```diff
@@ -188,21 +311,116 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+                        features_subset.append(features[:, random_indices[k]])
+                        feature_list.append(features)
+
+                    features_subset = torch.cat(features_subset, dim=1)
+                    feature_list = torch.cat(feature_list, dim=1)
+                    output = soft_assignments(feature_list, cluster_centers)
+                    target = target_distribution(output).detach()
```
Comment (pollytur) [inline sensorium/training/trainers.py]: again, don't do `detach()` before you did `loss.backwards()` (this thing actually precomputes the gradients and then you can detach)

File: sensorium/training/trainers.py [hunk 13]
```diff
@@ -188,21 +311,116 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+                    """
```
Comment (pollytur) [inline sensorium/training/trainers.py]: what is this comment here and why is it here?

File: sensorium/training/trainers.py [hunk 14]
```diff
@@ -188,21 +311,116 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+                        # Determine the required shift
+                        shift = max(0, -min_target.item(), -min_output.item()) + 1e-5  # A small epsilon to avoid zeros
+                        shifted_targets =target + shift
+                        shifted_output = output + shift
+                        target = shifted_targets / shifted_targets.sum(dim=-1, keepdim=True)
+                        output = shifted_output / shifted_output.sum(dim=-1, keepdim=True)
+                    """
+                    kldiv_loss = (
+                        get_multiplier(epoch)
+                        * kldiv_criterion(output.log(), target)
```
Comment (pollytur) [inline sensorium/training/trainers.py:358]: add a small comment with torch docs or some other refs, eg why its `outputs.log()` here (I know we had this discussion but it might potentially help someone if someone else is going to work with your code afterwards)

File: sensorium/training/trainers.py [hunk 15]
```diff
@@ -168,6 +252,40 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+            model.train()
+            feature_list = []
+            features_subset = []
+            random_indices = {}
+            # form initial cluster centres
+
+            for k, readout in model.readout.items():
+                features = readout.features.cpu().detach().squeeze().T.numpy()
+                feature_list.append(np.array(features))
+                random_index = np.random.choice(
```
Comment (pollytur) [inline sensorium/training/trainers.py]: this is something that I mentioned before - if you run this code twice - you might get different `random_index`. So (1)fix a seed here as well, (2) give it an option to read `random_index` dictionary from some pre-saved file to have it same across runs

File: sensorium/training/trainers.py (hunk 9 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: True

File: sensorium/training/trainers.py (hunk 13 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: That is for the case that one of the distributions is less than 0. Then the KL loss isn't defined. But now I'm not using it

Comment (pollytur) [reply sensorium/training/trainers.py]: do you think we still need it or can we delete it? (both ways is fine)

File: sensorium/training/trainers.py (hunk 10 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: True. Thank you

File: sensorium/training/trainers.py (hunk 11 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: I changed that

File: sensorium/training/trainers.py (hunk 15 above)
Comment (Nisone2000) [reply sensorium/training/trainers.py]: I will do that. I guess I will remove the subsamples at the end because the training runs through with the whole feature set. It was more to speed up the debugging process

Comment (pollytur) [reply sensorium/training/trainers.py]: yes, for that sake of making the first things run - this is totally fine:) maybe just add a todo for yourself in the code that we don't forget it later

File: sensorium/training/trainers.py (hunk 11 above)
Comment (pollytur) [reply sensorium/training/trainers.py]: if you commit and push to the same branch (`[KLdiv_branch`) I should see the changes here (the code in PR is automatically updated once you push to the source branch)

Comment (pollytur) [reply sensorium/training/trainers.py]: but better do it after addressing all the comments:)

File: sensorium/training/trainers.py [hunk 16]
```diff
@@ -168,6 +260,43 @@ def full_objective(model, dataloader, data_key, *args, **kwargs):
...
+                features = np.vstack(feature_list)
+                features_subset = np.vstack(features_subset)
+                predicted = kmeans.fit_predict(features)
+
+                # predicted_previous = torch.tensor(np.copy(predicted), dtype=torch.long)
+                # _, accuracy = cluster_accuracy(predicted, actual.cpu().numpy())
+                cluster_centers = torch.tensor(
+                    kmeans.cluster_centers_, dtype=torch.float, requires_grad=True
+                )
+                cluster_centers = cluster_centers.to(device, non_blocking=True)
```
Comment (pollytur) [inline sensorium/training/trainers.py:292]: I am not sure if these lines should be inside `with torch.no_grad():`, I actually think they should be outside like 
```
with torch.no_grad():
    for k, readout in model.readout.items():
        features = readout.features.cpu().detach().squeeze().T.numpy()
        ....
cluster_centers = torch.tensor(
    kmeans.cluster_centers_, dtype=torch.float, requires_grad=True
)
cluster_centers = cluster_centers.to(device, non_blocking=True)

```
